hr_commission/models/res_partner.py

```diff
@@ -1,55 +1,35 @@
-# -*- coding: utf-8 -*-
-##############################################################################
-#
-#    OpenERP, Open Source Management Solution
-#    Copyright (C) 2015 Avanzosc (<http://www.avanzosc.es>)
-#    Copyright (C) 2015 Pedro M. Baeza (<http://www.serviciosbaeza.com>)
-#
-#    This program is free software: you can redistribute it and/or modify
-#    it under the terms of the GNU General Public License as published by
-#    the Free Software Foundation, either version 3 of the License, or
-#    (at your option) any later version.
-#
-#    This program is distributed in the hope that it will be useful,
-#    but WITHOUT ANY WARRANTY; without even the implied warranty of
-#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
-#    GNU General Public License for more details.
-#
-#    You should have received a copy of the GNU General Public License
-#    along with this program.  If not, see <http://www.gnu.org/licenses/>.
-#
-##############################################################################
-from openerp import models, fields, api, exceptions, _
+# Copyright 2015-2018 Tecnativa - Pedro M. Baeza
+# License AGPL-3 - See https://www.gnu.org/licenses/agpl-3.0.html
+
+from odoo import _, api, fields, exceptions, models
 
 
 class ResPartner(models.Model):
-    """Add some fields related to commissions"""
     _inherit = "res.partner"
 
     agent_type = fields.Selection(
-        selection_add=[("salesman", "Salesman (employee)")])
-    employee = fields.Many2one(
-        comodel_name="hr.employee", compute="_get_employee")
-    users = fields.One2many(comodel_name="res.users",
-                            inverse_name="partner_id")
+        selection_add=[("salesman", "Salesman (employee)")],
+    )
+    employee_id = fields.Many2one(
+        comodel_name="hr.employee", compute="_compute_employee_id",
+    )
 
-    @api.one
-    @api.depends('users')
-    def _get_employee(self):
-        self.employee = False
-        if len(self.users) == 1 and len(self.users[0].employee_ids) == 1:
-            self.employee = self.users[0].employee_ids[0]
+    @api.depends('user_ids')
+    def _compute_employee_id(self):
+        for partner in self:
+            if (len(partner.user_ids) == 1 and
+                    len(partner.user_ids[0].employee_ids) == 1):
+                partner.employee_id = partner.user_ids[0].employee_ids[0]
 
-    @api.constrains('agent_type', 'employee')
+    @api.constrains('agent_type')
```

Comment:
@pedrobaeza All field used by the constrains must be specified into `@api.constrains` to trigger the constrains check if value change. `employee_id` must be in `@api.constrains

Reply:
As that field is a computed non-stored value, it shouldn't be there, and a warning reminds you about that. This is how I saw it.

Reply:
Therefore you should add the field declare into the depends of the compute method...

Reply:
But the compute is precisely for computing that field...

Reply:
The code should look like:

```python
   @api.depends('user_ids')
    def _compute_employee_id(self):
        for partner in self:
            if (len(partner.user_ids) == 1 and
                    len(partner.user_ids.employee_ids) == 1):
                partner.employee_id = partner.user_ids.employee_ids

    @api.constrains('agent_type', 'user_ids)
    def _check_employee(self):
        if self.agent_type == 'salesman' and not self.employee_id:
            raise exceptions.ValidationError(
                _("To be salesman there must be one (and only one) employee linked to this "
                  "partner. To do this, go to 'Human Resources' and check "
                  "'Employees'"))
```

If `user_ids` is not into `api.constrains` and you remove/modify the link between `res.parner` and `hr.employee` the constrains will not be triggered.

Reply:
Thanks for pointer. Now I know what you are talking about, but are you sure adding the one2many field to the constraint, it will be triggered on that case? I can develop a test case any way.

Reply:
I have tried and it's not triggered, so I'm not sure about adding a lot of code in employee part just for that corner case. Do you see it so important to do it?

Reply:
IMO, yes it's important to avoid know bug.